### PR TITLE
Photography

### PR DESCRIPTION
**Pull request by RemieRichards**
Cameras, photos and camera film.

Photos are viewable by anyone who picks up and opens the photo, clients don't download the photo until they open the photo itself so creating photos isn't a burden, photos are only sent to each client once.

https://imgur.com/y8JlF6n

### CODE REVIEW

File: Content.Server/GameObjects/Components/Photography/PhotoComponent.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        public bool UseEntity(UseEntityEventArgs eventArgs)
+        {
+            if(_path == null)
+            {
+                eventArgs.User.PopupMessage(eventArgs.User, "It's not dry yet!");
+                return false;
+            }
+
+            if (!_cached)
```
Comment (RemieRichards) [inline Content.Server/GameObjects/Components/Photography/PhotoComponent.cs:50]: just realised this logic is a bit wrong.
I get what I was going for but this currently works for 1 player only.

Will fix when I get home in 5~ hours.

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        [ViewVariables] public bool CameraOn { get; private set; } = false;
+        [ViewVariables] public int Radius { get; private set; } = 0;
+        [ViewVariables] public int Film { get; private set; } = 0;
+        [ViewVariables] public int FilmMax { get; private set; } = 0;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            IoCManager.InjectDependencies(this);
```
Comment (ComicIronic) [inline Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:51]: Why is this necessary?

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            var cropDimensions = EyeManager.PixelsPerMeter * (Radius * 4);
+
+            //We'll try and center, but otherwise we'll shift the box so it doesn't go outside the viewport
+            var cropX = (int)Math.Clamp(Math.Floor(photoCenter.X - cropDimensions / 2), 0, screenshot.Width - cropDimensions);
+            var cropY = (int)Math.Clamp(Math.Floor(photoCenter.Y - cropDimensions / 2), 0, screenshot.Height - cropDimensions);
+
+            Logger.InfoS("photo", $"cropX:{cropX}, cropY:{cropY}, w:{screenshot.Width}, h:{screenshot.Height}");
+
+            var success = false;
+            var filename = $"MY_PHOTO";
```
Comment (ComicIronic) [inline Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:154]: ?

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+            using (screenshot)
+            {
+                screenshot.Mutate(
+                    i => i.Crop(new Rectangle(cropX, cropY, cropDimensions, cropDimensions))
+                );
+
+                for (var i = 0; i < 5; i++)
```
Comment (ComicIronic) [inline Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:161]: Does this only support 5 photo files?

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                            screenshot.SaveAsPng(file);
+                        });
+
+                        success = true;
+                        Logger.InfoS("photo", "Photo taken as {0}.png", filename);
+                        break;
+                    }
+                    catch (IOException e)
+                    {
+                        Logger.WarningS("photo", "Failed to save photo, retrying?:\n{0}", e);
```
Comment (ComicIronic) [inline Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:184]: I don't know if this will dump the whole stack trace in logs or not. Might be nicer to just use `e.Message` if it does result in the full trace.

File: Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            IoCManager.InjectDependencies(this);
+            _inputSystem = EntitySystemManager.GetEntitySystem<InputSystem>();
+        }
+
+        public override void Update(float frameTime)
```
Comment (ComicIronic) [inline Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs:34]: This seems like a strangely unnecessary approach to what should be a RangedAttack (or RangedInteract I believe it's called now). Was there a problem doing it that way?

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 3 above)
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:154]: this and the next 3 comments are just identical code from taking a screenshot.
Because that's all this does, is take a screenshot, crop it, and then try and save it.

The file being saved here is the last photo you took, it's saved to disk as a PNG, and later that same PNG file's bytes are uploaded to the server (so they can be reinterpreted and resaved as a PNG on the server side, ready to serve up the photo to anyone who looks at it)

File: Content.Server/GameObjects/Components/GUI/InventoryComponent.cs [hunk 7]
```diff
@@ -175,9 +175,9 @@ public bool CanEquip(Slots slot, ItemComponent item, out string reason)
         /// </summary>
         /// <param name="slot">The slot to drop the item from.</param>
         /// <returns>True if an item was dropped, false otherwise.</returns>
-        public bool Unequip(Slots slot)
+        public bool Unequip(Slots slot, bool force = false)
```
Comment (ComicIronic) [inline Content.Server/GameObjects/Components/GUI/InventoryComponent.cs:178]: New params could use doc comments.

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 4 above)
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:161]: This is 5 attempts to save the same file.
Exact same precautions PJ took with screenshots.
In practice I've only noticed it get as bad as the third attempt to save it (dash 2).

Since this is just the photo you last took, there's no need to keep it around forever, so they get overwritten if you take another photo.

(Ideally on client shutdown or somewhere we'd then delete all of these (probably need to make a /Photos directory like there's a /Screenshots one, just for organisation))

File: Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+            }
+            else
+            {
+                message.AddMarkup(Loc.GetString("The {0} is [color=red]Off[/color]", Owner.Name));
+            }
+
+            if (inDetailsRange)
+            {
+                message.AddMarkup(Loc.GetString("\nFilm: [color={0}]{1}/{2}[/color], ",
+                    Film<=0 ? "red" : "white", Film, FilmMax));
```
Comment (ComicIronic) [inline Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs:221]: No `HasFilm`?

File: Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs (hunk 6 above)
Comment (RemieRichards) [reply Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs:34]: If you can point me towards the classes and files I'd be happy to rework it, as it stands this is the same approach guns do (Perhaps guns need the same refactor, or have some extra conditions that require this code)

File: Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 8 above)
Comment (RemieRichards) [reply Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs:221]: Added `HasFilm` later on, can certainly replace the `FIlm<=0` with it.

File: Content.Server/GameObjects/Components/GUI/InventoryComponent.cs (hunk 7 above)
Comment (RemieRichards) [reply Content.Server/GameObjects/Components/GUI/InventoryComponent.cs:178]: Will do a doc pass at some point, probably tonight since I'll be fixing some stuff (and maybe the refactor suggested above too) anyway.

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 5 above)
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:184]: Never actually had this occur, but I'll force one and see how it looks.

File: Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 2 above)
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoCameraComponent.cs:51]: Looks like this isn't anymore. oops.

**Comment by RemieRichards** [comment]
Thanks @ComicIronic 
(also, is that everything or just everything for now?)

File: Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs (hunk 6 above)
Comment (ComicIronic) [reply Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs:34]: (I made a mistake, it's actually `IAfterInteract` that's necessary)

`IAfterInteract` on the server component should give click behaviour at range when no interaction happens, which could be used for the camera.
The downsides I can see are:
  1. There's a message shuffle - click goes to the server, then the server tells the client to take the photograph, then the client sends the photo back.
  2. Maybe there are some non-obvious behaviour differences?

The main reason I can think of that guns would use special code is that they are old and they need special handling for full-auto when you hold down the mouse. @PJB3005 You wrote it 2 years ago, any insights?

**Comment by ComicIronic** [comment]
I think that's everything, but sometimes I re-review things and spot new stuff. I looked at all the files, at least.

File: Content.Server/GameObjects/Components/GUI/InventoryComponent.cs [hunk 9]
```diff
@@ -206,15 +206,27 @@ public bool Unequip(Slots slot)
         /// <returns>
         ///     True if there is an item in the slot and it can be dropped, false otherwise.
         /// </returns>
-        public bool CanUnequip(Slots slot)
+        public bool CanUnequip(Slots slot, bool force = false)
         {
-            if (!ActionBlockerSystem.CanUnequip(Owner))
+            if (!ActionBlockerSystem.CanUnequip(Owner) && !force)
```
Comment (RemieRichards) [inline Content.Server/GameObjects/Components/GUI/InventoryComponent.cs:211]: I'll swap this around as well.

File: Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+                        }
+                    }
+
+                    //Store photo to disk, assign path to photo entity
+                    _photoSystem.StorePhoto(photoTaken.Data, photo);
+
+                    break;
+            }
+        }
+
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs:174]: Many security holes here:

1. It trusts the client on the author of the photo (can make photos appear in anybody's hands)
2. Doesn't verify that the client is alive/has the camera/camera is on/camera has film.
3. Film and the sound are separate from the actual action of taking the photo meaning that a hacked client could trivially circumvent the film requirement.

A less trivial issue is that the image sent by the client could very easily be malicious. An example is that the size could be something like 100000x100000 of whitespace which could easily crash clients when they try to load it, but compress to "reasonable" sizes. You'll probably have to make the server load the PNG in some sort of safe way to reject it if malicious (hopefully ImageSharp allows inspecting the size without loading it?).

File: Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+        {
+            protected override void GetData(IEntity user, PhotoCameraComponent component, VerbData data)
+            {
+                data.CategoryData = VerbCategories.PhotoRadius;
+                data.Text = "2x2";
+            }
+
+            protected override void Activate(IEntity user, PhotoCameraComponent component)
+            {
+                component.Radius = 1;
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs:269]: These verbs need to check whether the user is actually holding the camera and not incapacitated or such.

File: Content.Client/GameObjects/Components/Items/ClientHandsComponent.cs [hunk 12]
```diff
@@ -59,6 +59,7 @@ public override void Initialize()
 
         public IEntity GetEntity(string index)
         {
+            if (index == null) return null;
```
[the remark this answers is not in the document]
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Items/ClientHandsComponent.cs:62]: This happened as part of the client-side PhotoSystem's update() loop, somehow it was passing along null for the active hand string.

File: Content.Client/GameObjects/Components/Photography/PhotoWindow.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+using Robust.Shared.Maths;
+using System.IO;
+
+namespace Content.Client.GameObjects.Components
+{
+    public class PhotoWindow : SS14Window
+    {
+        private PhotoSystem _photoSystem;
+
+        protected override Vector2? CustomSize => (0,0); //texture rect fills the ui and scales it for us
```
[the remark this answers is not in the document]
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoWindow.cs:15]: Fair point, I'll double check tonight.

File: Content.Client/GameObjects/Components/Photography/PhotoWindow.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        private TextureRect _photo;
+
+        public PhotoWindow()
+        {
+            Title = "Photo";
```
[the remark this answers is not in the document]
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoWindow.cs:20]: Will do a localisation pass for any places I missed.

File: Content.Client/GameObjects/Components/Photography/PhotoWindow.cs [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            container.AddChild(_photo);
+            Contents.AddChild(container);
+
+            _photoSystem = EntitySystem.Get<PhotoSystem>();
+        }
+
+        public async void Populate(string photoId)
+        {
+            var photoBytes = await _photoSystem.GetPhotoBytes(photoId);
+            _photo.Texture = Texture.LoadFromPNGStream(new MemoryStream(photoBytes));
```
[the remark this answers is not in the document]
Comment (RemieRichards) [reply Content.Client/GameObjects/Components/Photography/PhotoWindow.cs:32]: Sounds good.

File: Content.Client/GameObjects/EntitySystems/PhotoSystem.cs [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+        private bool _blocked;
+        private bool _down;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            IoCManager.InjectDependencies(this);
+            _inputSystem = Get<InputSystem>();
+            _photoSystem = Get<PhotoSystem>();
```
[the remark this answers is not in the document]
Comment (RemieRichards) [reply Content.Client/GameObjects/EntitySystems/PhotoSystem.cs:55]: Yeah I appear to have temporarily gone brain dead while writing that part (I think it may have moved between classes a few times, before I solidified on having PhotoSystem(s) handle all image stuff)

File: Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 11 above)
Comment (RemieRichards) [reply Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs:269]: D'oh.

File: Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs (hunk 10 above)
Comment (RemieRichards) [reply Content.Server/GameObjects/Components/Photography/PhotoCameraComponent.cs:174]: I'll Just spawn the photo where the camera is, and if the camera is being held, put it in that entity's hands.
(Ditto for suicides (also also, should I just cut that? it's mildly funny but it shits up the code a bit))

Put some messages in disc about the image bombs, had the same countermeasure thoughts.

File: Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs (hunk 6 above)
Comment (RemieRichards) [reply Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs]: @PJB3005 would you recommend leaving this as is for now then? (to avoid the issues you mentioned)

**Comment by RemieRichards** [comment]
@PJB3005 thanks for the review, will fix it tonight.

File: Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs (hunk 6 above)
Comment (PJB3005) [reply Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs]: You still don't need an update method. Now with chairbender's input binding rework you can just bind the Use key function going down instead of using an update loop like this.

Comment (RemieRichards) [reply Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs]: I thought the input rework wasn't merged yet? engine 1132?

Comment (RemieRichards) [reply Content.Client/GameObjects/EntitySystems/ClientPhotoSystem.cs]: Ah no I see now, will do.

**Comment by SweptWasTaken** [comment]
>  review, will fix it tonight.

long night

**Comment by RemieRichards** [comment]
I exist outside of space and time.
no bully.

**Comment by PJB3005** [comment]
@SweptWasTaken you still haven't finished eating that pizza.

**Comment by SweptWasTaken** [comment]
image
